File: hackerrank/30 days of code/day 23, bst level-order traversal/bst_level_order_traversal.cs
// https://www.hackerrank.com/challenges/30-binary-trees/problem?isFullScreen=true

using static System.Console;

record Node(int Data)
{
    public Node? Left { get; set; }
    public Node? Right { get; set; }
}

static class Solution
{
    static void Main()
    {
        Node? root = ReadTree();
        LevelOrder(root);
    }

    private static Node? ReadTree()
    {
        Node? root = null;
        for (int n = int.Parse(ReadLine()); n-- > 0;)
        {
            int data = int.Parse(ReadLine());
            root = InsertNode(root, data);
        }
        return root;
    }

    private static Node InsertNode(Node? root, int data)
    {
        if (root == null)
            return new Node(data);
        if (data <= root.Data)
            root.Left = InsertNode(root.Left, data);
        else
            root.Right = InsertNode(root.Right, data);
        return root;
    }

    private static void LevelOrder(Node? root)
    {
        if (root == null)
            return;

        var myQueue = new Queue<Node>();
        myQueue.Enqueue(root);
        while (myQueue.Count != 0)
        {
            var current = myQueue.Dequeue();
            Write(current.Data + " ");
            if (current.Left != null)
                myQueue.Enqueue(current.Left);
            if (current.Right != null)
                myQueue.Enqueue(current.Right);
        }
    }
}

File: hackerrank/30 days of code/day 23, bst level-order traversal/bst_level_order_traversal.js
// https://www.hackerrank.com/challenges/30-binary-trees/problem?isFullScreen=true

'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let inputLines = [];
let currentLine = 0;

process.stdin.on('data', function (inputStdin) {
  inputString += inputStdin;
});

process.stdin.on('end', function () {
  inputLines = inputString.split('\n');
  inputString = '';
  main();
});

function readLine() {
  return inputLines[currentLine++];
}

//////////////////////////////////////////////////

function main() {
  const root = readTree();
  levelOrder(root);
}

function readTree() {
  let root = null;
  for (let n = +readLine(); n-- > 0;) {
    const data = +readLine();
    root = insertNode(root, data);
  }
  return root;
}

function insertNode(root, data) {
  if (root == null)
    return new Node(data);
  if (data <= root.data)
    root.left = insertNode(root.left, data);
  else
    root.right = insertNode(root.right, data);
  return root;
}

class Node {
  constructor(data) {
    this.data = data;
    this.left = null;
    this.right = null;
  }
}

function levelOrder(root) {
  if (!root) return;

  let myQueue = [];
  myQueue.push(root);
  for (let i = 0; i < myQueue.length; i++) {
    process.stdout.write(myQueue[i].data + ' ');
    if (myQueue[i].left)
      myQueue.push(myQueue[i].left);
    if (myQueue[i].right)
      myQueue.push(myQueue[i].right);
  }
}
